File: app/src/main/java/com/bopr/android/smailer/ui/RulesFragment.java
package com.bopr.android.smailer.ui;

import android.content.Intent;
import android.os.Bundle;

import androidx.preference.Preference;
import androidx.preference.Preference.OnPreferenceChangeListener;

import com.bopr.android.smailer.R;
import com.bopr.android.smailer.util.TagFormatter.TagPattern;

import java.util.Set;

import static com.bopr.android.smailer.Settings.PREF_EMAIL_TRIGGERS;
import static com.bopr.android.smailer.Settings.PREF_FILTER_PHONE_BLACKLIST;
import static com.bopr.android.smailer.Settings.PREF_FILTER_PHONE_WHITELIST;
import static com.bopr.android.smailer.Settings.PREF_FILTER_TEXT_BLACKLIST;
import static com.bopr.android.smailer.Settings.PREF_FILTER_TEXT_WHITELIST;
import static com.bopr.android.smailer.util.TagFormatter.formatter;
import static com.bopr.android.smailer.util.TextUtil.commaSplit;

/**
 * Conditions settings activity's fragment.
 *
 * @author Boris Pronin (<a href="mailto:boprsoft.dev@gmail.com">boprsoft.dev@gmail.com</a>)
 */
public class RulesFragment extends BasePreferenceFragment {

    @Override
    public void onCreatePreferences(Bundle bundle, String rootKey) {
        addPreferencesFromResource(R.xml.pref_rules);

        requirePreference(PREF_EMAIL_TRIGGERS).setOnPreferenceChangeListener(new OnPreferenceChangeListener() {

            @Override
            public boolean onPreferenceChange(Preference preference, Object value) {
                if (value == null || ((Set) value).isEmpty()) {
                    updateSummary(preference, getString(R.string.no_triggers_specified), SUMMARY_STYLE_ACCENTED);
                } else {
                    updateSummary(preference, getString(R.string.events_causing_sending_mail), SUMMARY_STYLE_DEFAULT);
                }
                return true;
            }
        });

        Preference phoneBlacklistPreference = requirePreference(PREF_FILTER_PHONE_BLACKLIST);
        phoneBlacklistPreference.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {

            @Override
            public boolean onPreferenceClick(Preference preference) {
                startActivity(new Intent(getContext(), PhoneBlacklistActivity.class));
                return true;
            }
        });
        phoneBlacklistPreference.setOnPreferenceChangeListener(new OnPreferenceChangeListener() {

            @Override
            public boolean onPreferenceChange(Preference preference, Object value) {
                updateSummary(preference, formatListAndSize((String) value, R.string.unacceptable_phone_numbers,
                        R.string._none), SUMMARY_STYLE_DEFAULT);
                return true;
            }
        });

        Preference phoneWhitelistPreference = requirePreference(PREF_FILTER_PHONE_WHITELIST);
        phoneWhitelistPreference.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {

            @Override
            public boolean onPreferenceClick(Preference preference) {
                startActivity(new Intent(getContext(), PhoneWhitelistActivity.class));
                return true;
            }
        });
        phoneWhitelistPreference.setOnPreferenceChangeListener(new OnPreferenceChangeListener() {

            @Override
            public boolean onPreferenceChange(Preference preference, Object value) {
                updateSummary(preference, formatListAndSize((String) value, R.string.acceptable_phone_numbers,
                        R.string._any), SUMMARY_STYLE_DEFAULT);
                return true;
            }
        });

        Preference textBlacklistPreference = requirePreference(PREF_FILTER_TEXT_BLACKLIST);
        textBlacklistPreference.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {

            @Override
            public boolean onPreferenceClick(Preference preference) {
                startActivity(new Intent(getContext(), TextBlacklistActivity.class));
                return true;
            }
        });
        textBlacklistPreference.setOnPreferenceChangeListener(new OnPreferenceChangeListener() {

            @Override
            public boolean onPreferenceChange(Preference preference, Object value) {
                updateSummary(preference, formatListAndSize((String) value, R.string.unacceptable_words,
                        R.string._none), SUMMARY_STYLE_DEFAULT);
                return true;
            }
        });

        Preference textWhitelistPreference = requirePreference(PREF_FILTER_TEXT_WHITELIST);
        textWhitelistPreference.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {

            @Override
            public boolean onPreferenceClick(Preference preference) {
                startActivity(new Intent(getContext(), TextWhitelistActivity.class));
                return true;
            }
        });
        textWhitelistPreference.setOnPreferenceChangeListener(new OnPreferenceChangeListener() {

            @Override
            public boolean onPreferenceChange(Preference preference, Object value) {
                updateSummary(preference, formatListAndSize((String) value, R.string.acceptable_words,
                        R.string._any), SUMMARY_STYLE_DEFAULT);
                return true;
            }
        });

    }

    private String formatListAndSize(String value, int patternRes, int emptyTextRes) {
        TagPattern pattern = formatter(requireContext()).pattern(patternRes);
        int size = commaSplit(value).size();
        if (size > 0) {
            pattern.put("size", String.valueOf(size));
        } else {
            pattern.put("size", emptyTextRes);
        }
        return pattern.format();
    }

}

File: app/src/main/kotlin/com/bopr/android/smailer/ui/RulesFragment.kt
package com.bopr.android.smailer.ui

import android.content.BroadcastReceiver
import android.os.Bundle
import androidx.annotation.StringRes
import com.bopr.android.smailer.R
import com.bopr.android.smailer.Settings.Companion.PREF_MAIL_TRIGGERS
import com.bopr.android.smailer.data.Database
import com.bopr.android.smailer.data.Database.Companion.TABLE_PHONE_BLACKLIST
import com.bopr.android.smailer.data.Database.Companion.TABLE_PHONE_WHITELIST
import com.bopr.android.smailer.data.Database.Companion.TABLE_TEXT_BLACKLIST
import com.bopr.android.smailer.data.Database.Companion.TABLE_TEXT_WHITELIST
import com.bopr.android.smailer.data.Database.Companion.registerDatabaseListener
import com.bopr.android.smailer.data.Database.Companion.unregisterDatabaseListener
import com.bopr.android.smailer.util.SummaryStyle.SUMMARY_STYLE_ACCENTED
import com.bopr.android.smailer.util.requirePreference
import com.bopr.android.smailer.util.setOnChangeListener
import com.bopr.android.smailer.util.updateSummary

/**
 * Rules settings fragment.
 *
 * @author Boris Pronin ([boprsoft.dev@gmail.com](mailto:boprsoft.dev@gmail.com))
 */
class RulesFragment : BasePreferenceFragment(R.xml.pref_rules) {

    private lateinit var database: Database
    private lateinit var databaseListener: BroadcastReceiver

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        requirePreference(PREF_MAIL_TRIGGERS).setOnChangeListener {
            it.apply {
                if (settings.getStringSet(key).isEmpty()) {
                    updateSummary(R.string.no_triggers_specified, SUMMARY_STYLE_ACCENTED)
                } else {
                    updateSummary(R.string.email_processing_triggers)
                }
            }
        }

        database = Database(requireContext())
        databaseListener = requireContext().registerDatabaseListener { tables ->
            if (tables.contains(TABLE_PHONE_BLACKLIST)) updatePhoneBlacklistPreferenceView()
            if (tables.contains(TABLE_PHONE_WHITELIST)) updatePhoneWhitelistPreferenceView()
            if (tables.contains(TABLE_TEXT_BLACKLIST)) updateTextBlacklistPreferenceView()
            if (tables.contains(TABLE_TEXT_WHITELIST)) updateTextWhitelistPreferenceView()
        }
    }

    override fun onDestroy() {
        requireContext().unregisterDatabaseListener(databaseListener)
        database.close()

        super.onDestroy()
    }

    override fun onStart() {
        super.onStart()

        updatePhoneBlacklistPreferenceView()
        updatePhoneWhitelistPreferenceView()
        updateTextBlacklistPreferenceView()
        updateTextWhitelistPreferenceView()
    }

    private fun updatePhoneBlacklistPreferenceView() {
        requirePreference(PREF_PHONE_BLACKLIST).updateSummary(
            formatListSummary(
                database.phoneBlacklist, R.string.unacceptable_phone_numbers, R.string._none
            )
        )
    }

    private fun updatePhoneWhitelistPreferenceView() {
        requirePreference(PREF_PHONE_WHITELIST).updateSummary(
            formatListSummary(
                database.phoneWhitelist, R.string.acceptable_phone_numbers, R.string._any
            )
        )
    }

    private fun updateTextBlacklistPreferenceView() {
        requirePreference(PREF_TEXT_BLACKLIST).updateSummary(
            formatListSummary(
                database.textBlacklist, R.string.unacceptable_words, R.string._none
            )
        )
    }

    private fun updateTextWhitelistPreferenceView() {
        requirePreference(PREF_TEXT_WHITELIST).updateSummary(
            formatListSummary(
                database.textWhitelist, R.string.acceptable_words, R.string._any
            )
        )
    }

    private fun formatListSummary(
        list: Set<String>, @StringRes patternRes: Int,
        @StringRes emptyRes: Int
    ): String {
        return getString(patternRes, if (list.isEmpty()) getString(emptyRes) else list.size)
    }

    companion object {

        private const val PREF_TEXT_WHITELIST = "text_whitelist"
        private const val PREF_TEXT_BLACKLIST = "text_blacklist"
        private const val PREF_PHONE_WHITELIST = "phone_whitelist"
        private const val PREF_PHONE_BLACKLIST = "phone_blacklist"
    }
}